node/src/components/consensus/era_supervisor.rs

```diff
@@ -1329,3 +1319,17 @@ where
         valid,
     }))
 }
+
+impl ProposedBlock<ClContext> {
+    /// If this block contains a deploy that's also present in an ancestor, this returns the deploy
+    /// hash, otherwise `None`.
+    fn contains_replay(&self) -> Option<&DeployHash> {
+        let block_deploys_set: BTreeSet<DeployHash> =
+            self.value().deploys_and_transfers_iter().cloned().collect();
+        self.context()
+            .ancestor_values()
+            .iter()
+            .flat_map(|ancestor| ancestor.deploys_and_transfers_iter())
+            .find(|deploy| block_deploys_set.contains(deploy))
```

Comment:
Only one thing I wonder – is this guaranteed to be the same `DeployHash` on every node that executes this method? Or is the order non-deterministic? I was imagining validators debugging this and _maybe_ comparing their logs.

Reply:
The order of iteration should be deterministic. In the block payload it's even `Vec`s, and the ancestors are sorted in reverse chronological order. So every validator will see the same message.